ladybug/header.py

```diff
@@ -1,103 +1,142 @@
-        self.middle_hour = middle_hour if middle_hour else False
+        data_type = data_type or 'Unknown Data'
+        self.set_data_type_and_unit(data_type, unit)
+        self.analysis_period = analysis_period
+        self.location = location
 
     @classmethod
     def from_json(cls, data):
         """Create a header from a dictionary.
 
         Args:
             data: {
-                "location": {}, // A Ladybug location
-                "data_type": string, //Type of data (e.g. Temperature) (Default: None).
+                "data_type": {}, //Type of data (e.g. Temperature)
                 "unit": string,
-                "analysis_period": {}, // A Ladybug AnalysisPeriod,
-                "middle_hour": {} // Whether values fall in the middle of the hour
+                "analysis_period": {} // A Ladybug AnalysisPeriod
+                "location": {}, // A Ladybug location
             }
         """
         # assign default values
-        keys = ('location', 'data_type', 'unit', 'analysis_period', 'middle_hour')
+        keys = ('data_type', 'unit', 'analysis_period', 'location')
         for key in keys:
             if key not in data:
                 data[key] = None
 
-        location = Location.from_json(data['location'])
+        data_type = DataTypeBase.from_json(data['data_type'])
         ap = AnalysisPeriod.from_json(data['analysis_period'])
-        return cls(location, data['data_type'], data['unit'], ap, data['middle_hour'])
+        location = Location.from_json(data['location'])
+        return cls(data_type, data['unit'], ap, location)
 
     @classmethod
     def from_header(cls, header):
         """Try to generate a header from a header or a header string."""
         if hasattr(header, 'isHeader'):
             return header
 
-        # "%s|%s(%s)|%s"
+        # "%s(%s)|%s|%s"
         try:
             _h = header.replace("|", "**").replace("(", "**").replace(")", "")
             return cls(*_h.split("**"))
         except Exception as e:
             raise ValueError(
                 "Failed to create a Header from %s!\n%s" % (header, e))
 
+    @property
+    def data_type(self):
+        """A DataType object."""
+        return self._data_type
+
+    @property
+    def unit(self):
+        """A text string representing an abbreviated unit."""
+        return self._unit
+
+    @property
+    def analysis_period(self):
+        """A AnalysisPeriod object."""
+        return self._analysis_period
+
+    @analysis_period.setter
+    def analysis_period(self, ap):
+        self._analysis_period = AnalysisPeriod.from_analysis_period(ap)
+
+    @property
+    def location(self):
+        """A Location object."""
+        return self._location
+
+    @location.setter
+    def location(self, loc):
+        self._location = Location.from_location(loc)
+
     @property
     def isHeader(self):
         """Return True."""
         return True
 
+    def set_data_type_and_unit(self, data_type, unit):
```

Comment:
Without this method, there is no way to change the data_type on a header to be a totally new type of data.  This is because, when you set the unit, it checks to be sure that the unit is compatible with the currently assigned data_type and, when  you set the data_type, it checks to make sure that the currently assigned unit is not incompatible compatible with that data_type.

So, let's say that I want to take an hourly DataCollection of Radiation in Wh/m2 and I want to change it to be a DataCollection of Irradiance in W/m2.  I would need to use this set_data_type_and_unit method to do so since setting them individually would raise and exception.